packages/sats-wagmi/src/hooks/useBalance.tsx

```diff
@@ -29,10 +30,49 @@ const useBalance = (props: UseBalanceProps = {}) => {
       }
 
       const esploraClient = new EsploraClient(network);
+      const ordinalsClient = new OrdinalsClient(network);
 
-      const { confirmed, unconfirmed, total } = await esploraClient.getBalance(address);
+      const addressInfo = getAddressInfo(address);
 
-      return { confirmed: BigInt(confirmed), unconfirmed: BigInt(unconfirmed), total: BigInt(total) };
+      if (addressInfo.type === AddressType.p2tr) {
+        const [outputs, cardinalOutputs] = await Promise.all([
+          esploraClient.getAddressUtxos(address),
+          // cardinal = return UTXOs not containing inscriptions or runes
+          ordinalsClient.getOutputsFromAddress(address, 'cardinal')
+        ]);
+
+        const cardinalOutputsSet = new Set(cardinalOutputs.map((output) => output.outpoint));
+
+        const total = outputs.reduce((acc, output) => {
+          if (cardinalOutputsSet.has(OutPoint.toString(output))) {
+            return acc + output.value;
+          }
+
+          return acc;
+        }, 0);
```

Comment:
Looks to me like we expect the cardinals to be a subset of the outputs reported by esplora, is it necessary to do this cross-check? It would only exclude something reported by the ordinals API that is not known to esplora right?

Reply:
1. right, the original suggestion from @sander2 is [here](https://github.com/bob-collective/sats-wagmi/pull/125#discussion_r1842751607)
2. right. it's done to use `esplora` results as the source of truth, we assume that esplora might index faster
also `ord` treats utxo in mempool as unspent